### PR TITLE
[SR SQL Planner Summer Camp][Feature] alter table statement support of add list partition

### PR DESCRIPTION
**Pull request by hiliuxg**
## What type of PR is this：
- [ ] bug
- [X] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

issue #4035 

#### Create Table for single column partition

```sql
create database example_db  ;
use example_db ;

CREATE TABLE t_recharge_detail_single(
    id bigint  ,
    user_id  bigint  ,
    recharge_money decimal(32,2) , 
    province varchar ,
    dt varchar
) ENGINE=OLAP
DUPLICATE KEY(id)
PARTITION BY LIST (province) (
   PARTITION p1 VALUES IN ("beijing","chongqing"),
   PARTITION p2 VALUES IN ("shanghai","tianjing")
)
DISTRIBUTED BY HASH(id) BUCKETS 1
PROPERTIES(
    "replication_num" = "1"
) ;

alter table t_recharge_detail_single add partition p3 values in ("shanxi","guangdong") ;

show partitions  from  t_recharge_detail_single  \G; 

show create table t_recharge_detail_single  ;

```

#### Result
image
image

#### Create Table for multi column partition  

```sql
CREATE TABLE t_recharge_detail_multi(
    id bigint  ,
    user_id  bigint  ,
    recharge_money decimal(32,2) , 
    province varchar ,
    dt varchar
) ENGINE=OLAP
DUPLICATE KEY(id)
PARTITION BY LIST (dt,province) (
   PARTITION p1 VALUES IN (("2022-04-01", "beijing"),("2022-04-01", "chongqing")),
   PARTITION p2 VALUES IN (("2022-04-01", "shanghai"))
)
DISTRIBUTED BY HASH(id) BUCKETS 1
PROPERTIES(
    "replication_num" = "1"
);

alter table t_recharge_detail_multi add partition p3 values in (("2022-04-01", "shanxi"),("2022-04-01", "neimenggu"));
show partitions from t_recharge_detail_multi  \G;
show create table t_recharge_detail_multi  ;

```
#### Result
image
image

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 370 / 442 (83.71%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/PartitionDesc.java|0|8|00.00%|[59, 63, 67, 71, 75, 79, 83, 87]|
|:large_blue_circle:|com/starrocks/persist/RangePartitionPersistInfo.java|0|13|00.00%|[34, 35, 36, 39, 44, 45, 46, 47, 51, 52, 53, 54, 55]|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|3|00.00%|[212, 213, 214]|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|2|8|25.00%|[189, 190, 191, 192, 193, 194]|
|:large_blue_circle:|com/starrocks/catalog/RangePartitionInfo.java|3|4|75.00%|[243]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|227|260|87.31%|[902, 917, 918, 931, 944, 945, 946, 947, 948, 949, 950, 994, 1000, 1001, 1005, 1006, 1012, 1030, 1104, 1110, 1111, 1139, 1149, 1161, 1162, 1192, 1283, 1293, 1294, 1296, 1297, 1298, 1300]|
|:large_blue_circle:|com/starrocks/persist/PartitionPersistInfoV2.java|22|25|88.00%|[46, 50, 58]|
|:large_blue_circle:|com/starrocks/catalog/ListPartitionInfo.java|58|61|95.08%|[301, 307, 308]|
|:large_blue_circle:|com/starrocks/catalog/CatalogUtils.java|39|41|95.12%|[130, 131]|
|:large_blue_circle:|com/starrocks/analysis/SingleItemListPartitionDesc.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/persist/gson/GsonUtils.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/persist/ListPartitionPersistInfo.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/analysis/MultiItemListPartitionDesc.java|3|3|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/persist/OperationType.java [hunk 1]
```diff
@@ -247,4 +247,7 @@ public class OperationType {
     // shard operate 10221 ~ 10240
     public static final short OP_ADD_UNUSED_SHARD = 10221;
     public static final short OP_DELETE_UNUSED_SHARD = 10222;
+
+    // new operator for add partition 10321 ~ 10340
+    public static final short OP_ADD_PARTITION_V2 = 10321;
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/persist/OperationType.java:252]:
```suggestion
    // new operator for add partition 10241 ~ 10260
    public static final short OP_ADD_PARTITION_V2 = 10241;
```

Comment (wyb) [reply fe/fe-core/src/main/java/com/starrocks/persist/OperationType.java:252]: Add comment: only used in list partition currently.

File: fe/fe-core/src/main/java/com/starrocks/catalog/ListPartitionInfo.java [hunk 2]
```diff
@@ -127,23 +162,20 @@ public void write(DataOutput out) throws IOException {
      */
     public static PartitionInfo read(DataInput in) throws IOException {
         String json = Text.readString(in);
-        return GsonUtils.GSON.fromJson(json, ListPartitionInfo.class);
-    }
-
-    @Override
-    public void gsonPostProcess() throws IOException {
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/catalog/ListPartitionInfo.java:134]: This overwrites the original code.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 363 / 435 (83.45%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/PartitionDesc.java|0|8|00.00%|[59, 63, 67, 71, 75, 79, 83, 87]|
|:large_blue_circle:|com/starrocks/persist/RangePartitionPersistInfo.java|0|13|00.00%|[34, 35, 36, 39, 44, 45, 46, 47, 51, 52, 53, 54, 55]|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|3|00.00%|[212, 213, 214]|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|2|8|25.00%|[189, 190, 191, 192, 193, 194]|
|:large_blue_circle:|com/starrocks/catalog/RangePartitionInfo.java|3|4|75.00%|[243]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|226|259|87.26%|[902, 917, 918, 931, 944, 945, 946, 947, 948, 949, 950, 994, 1000, 1001, 1005, 1006, 1012, 1030, 1104, 1110, 1111, 1139, 1149, 1161, 1162, 1192, 1283, 1293, 1294, 1296, 1297, 1298, 1300]|
|:large_blue_circle:|com/starrocks/persist/PartitionPersistInfoV2.java|22|25|88.00%|[46, 50, 58]|
|:large_blue_circle:|com/starrocks/catalog/ListPartitionInfo.java|52|55|94.55%|[304, 310, 311]|
|:large_blue_circle:|com/starrocks/catalog/CatalogUtils.java|39|41|95.12%|[130, 131]|
|:large_blue_circle:|com/starrocks/analysis/SingleItemListPartitionDesc.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/persist/gson/GsonUtils.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/persist/ListPartitionPersistInfo.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/analysis/MultiItemListPartitionDesc.java|3|3|100.00%|[]|

**Review by wyb — APPROVED** [review]

**Review by gengjun-git — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Pull request merged**